data/pagination.go

```diff
@@ -0,0 +1,171 @@
+
+	limit, err := strconv.Atoi(query.Get("limit"))
+	if err != nil {
+		log.Event(ctx, "unable to convert search limit to int - set to default "+defaultLimitStr, log.INFO)
+		query.Set("limit", defaultLimitStr)
+		limit = cfg.DefaultLimit
+	}
+
+	if limit < cfg.DefaultLimit {
+		query.Set("limit", strconv.Itoa(cfg.DefaultLimit))
+		limit = cfg.DefaultLimit
+	}
+
+	if limit > cfg.DefaultMaximumLimit {
+		query.Set("limit", strconv.Itoa(cfg.DefaultMaximumLimit))
+		limit = cfg.DefaultMaximumLimit
+	}
+
+	return limit
+}
+
+func getPageFromURLQuery(ctx context.Context, cfg *config.Config, query url.Values) int {
+	defaultPageStr := strconv.Itoa(cfg.DefaultPage)
+
+	page, err := strconv.Atoi(query.Get("page"))
+	if err != nil {
+		log.Event(ctx, "unable to convert search page to int - set to default "+defaultPageStr, log.INFO)
+		query.Set("page", defaultPageStr)
+		page = cfg.DefaultPage
+	}
+
+	if page < 1 {
+		log.Event(ctx, "page number is less than default - default to page "+defaultPageStr, log.INFO)
+		query.Set("page", defaultPageStr)
+		page = cfg.DefaultPage
+	}
+
+	return page
+}
+
+func getOffset(ctx context.Context, cfg *config.Config, page int, limit int) (offset int, err error) {
+
+	offset = (page - 1) * limit
+
+	// when the offset is negative due to negative current page number or limit
+	if offset < 0 {
+		log.Event(ctx, "offset less than 0 - defaulted to offset "+strconv.Itoa(cfg.DefaultOffset), log.INFO)
+		offset = cfg.DefaultOffset
+	}
+
+	// when the offset is too big due to big current page number and/or limit
+	if (offset - limit) > cfg.DefaultMaximumSearchResults {
+		err = errs.ErrInvalidPage
+		logData := log.Data{"current_page": page, "limit": limit}
+
+		log.Event(ctx, "offset is too big as large page and/or limit given", log.Error(err), log.ERROR, logData)
+
+		return cfg.DefaultOffset, err
+	}
+
+	return offset, nil
+}
+
+// GetTotalPages gets the total pages of the search results
+func GetTotalPages(limit int, count int) int {
+	return (count + limit - 1) / limit
+}
+
+// GetPagesToDisplay gets all the pages available for the search results
+func GetPagesToDisplay(validatedQueryParams SearchURLParams, totalPages int) []model.PageToDisplay {
+	var pagesToDisplay = make([]model.PageToDisplay, 0)
+
+	currentPage := validatedQueryParams.CurrentPage
+
+	startPage := getStartPage(currentPage, totalPages)
+
+	endPage := getEndPage(startPage, totalPages)
+
+	controllerQuery := createSearchControllerQuery(validatedQueryParams)
+	query := controllerQuery.Get("q")
+
+	for i := startPage; i <= endPage; i++ {
+		pagesToDisplay = append(pagesToDisplay, model.PageToDisplay{
+			PageNumber: i,
+			URL:        getPageURL(query, i, controllerQuery),
+		})
+	}
+
+	return pagesToDisplay
+}
+
+func getStartPage(currentPage int, totalPages int) int {
+	startPage := currentPage - 2
+
+	if currentPage <= 2 {
+		startPage = 1
+	} else if (currentPage == totalPages-1) || (currentPage == totalPages) {
+		startPage = totalPages - 4
+	}
```

Comment:
As someone who has never seen the code before I don't know what the values 2, 1, 4 mean in this context or why they have been chosen. If you create constants for these values with a meaningful name that describes their purpose it makes it clearer to the reader what they are/represent.

Reply:
Of course. That makes complete sense - the little things are really important and I appreciate that